### PR TITLE
For #12565  Implement the common part of search widget in Android Components

### LINKED ISSUE #12565: Implement the common part of search widget in Android Components
I should move and refactor the search widget from Fenix in AC . This implementation should support the extension of the widget in Fenix and Focus . This has appear based on this https://github.com/mozilla-mobile/focus-android/issues/7249#issuecomment-1195533956 
This is the task from Focus https://github.com/mozilla-mobile/focus-android/issues/7249

### PR DESCRIPTION
**Pull request by iorgamgabriel**
### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/main/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/main/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### GitHub Automation
Fixes #12565

### CODE REVIEW

File: components/feature/searchwidget/src/main/java/mozilla/components/feature/searchwidget/IntentUtils.kt
Comment (Mugurell) [inline components/feature/searchwidget/src/main/java/mozilla/components/feature/searchwidget/IntentUtils.kt:10]: This is already available in `support.utils` which can be added as a dependency to the new feature.

File: components/feature/searchwidget/src/main/res/xml/mozac_search_widget_info.xml
Comment (Mugurell) [inline components/feature/searchwidget/src/main/res/xml/mozac_search_widget_info.xml:6]: Don't think we need this in AC?
AC can be left to handle the system broadcasts about the widget but the widget configuration - done in xml will be left for each client to take care of.
This will also mean AC doesn't need to offer that `previewImage` (which now has the Fenix icon).

File: components/feature/searchwidget/src/main/res/drawable/mozac_ic_launcher_foreground.xml
Comment (Mugurell) [inline components/feature/searchwidget/src/main/res/drawable/mozac_ic_launcher_foreground.xml:1]: A-C should be "client agnostic" and not include the launcher icon of Fenix.
If this resource is really needed maybe we can set the source to be `mozac_ic_globe`.

**Review by Mugurell — COMMENTED** [review]
Just a quick driveby, this is looking great!

File: components/feature/searchwidget/src/main/java/mozilla/components/feature/searchwidget/IntentUtils.kt
Comment (iorgamgabriel) [reply components/feature/searchwidget/src/main/java/mozilla/components/feature/searchwidget/IntentUtils.kt]: Done

File: components/feature/searchwidget/src/main/res/xml/mozac_search_widget_info.xml
Comment (iorgamgabriel) [reply components/feature/searchwidget/src/main/res/xml/mozac_search_widget_info.xml]: Done

File: components/feature/searchwidget/src/main/res/drawable/mozac_ic_launcher_foreground.xml
Comment (iorgamgabriel) [reply components/feature/searchwidget/src/main/res/drawable/mozac_ic_launcher_foreground.xml]: I added mozac_search_widget_icon that is an empty file . This file will be overridden in Focus and Fenix with the specific app icon .

File: .buildconfig.yml
Comment (jonalmeida) [inline .buildconfig.yml:115]: Why did we want a separate component? Would the `feature-search` have been fine? Just curious. 🙂

File: components/feature/searchwidget/src/main/res/values/colors.xml
Comment (jonalmeida) [inline components/feature/searchwidget/src/main/res/values/colors.xml:6]: @gabrielluong are these the right colour resources to use nowadays?

File: components/feature/searchwidget/src/main/java/mozilla/components/feature/searchwidget/AppSearchWidgetProvider.kt
Comment (jonalmeida) [inline components/feature/searchwidget/src/main/java/mozilla/components/feature/searchwidget/AppSearchWidgetProvider.kt:188]: For `R.drawable.mozac_search_widget_icon`, since we want the icon to be customizable, you could try to use simple data classes to hold configurations that the app can set. Similar to how you have a `abstract fun` you can have `abstract val`. This way we can land this component without needing to wait for designs/product direction on what the icon/configuration should look like, we can do the final configuration setup in Focus' integration side since we know we roughly want the same widget in Focus/Fenix.

For example:

```kotlin
abstract class AppSearchWidgetProvider {
  abstract val config: SearchWidgetConfig
}

data class SearchWidgetConfig(
  val iconResource: Int
  // insert other properties here that you'd like to make configurable on the app side.
)
```

**Review by jonalmeida — COMMENTED** [review]
Looks really nice! Left a few comments below too.

File: components/feature/searchwidget/src/main/res/values/colors.xml
Comment (gabrielluong) [reply components/feature/searchwidget/src/main/res/values/colors.xml:6]: I think the colors look right on a first glance. 

However, @iorgamgabriel can we import ui-colors into this and avoid duplicating the colors? We should only have a single source of truth for these color variables so that we can just change it in one place and get the expected changes. The problem right now is that the naming conventions of these colors are different from how they are defined by the design system - so if the design system changes layer color 1, we would not get the expected change for this module.

File: components/feature/searchwidget/src/main/res/values/strings.xml
Comment (gabrielluong) [inline components/feature/searchwidget/src/main/res/values/strings.xml:1]: Needs the license

File: components/feature/searchwidget/src/main/res/values/colors.xml
Comment (iorgamgabriel) [reply components/feature/searchwidget/src/main/res/values/colors.xml:6]: I imported ui-colors this is how i use photonLightGrey10 . I was thinking that search-widget should  have separate id for colors that could be overridden in client app (Focus and Fenix).

Comment (gabrielluong) [reply components/feature/searchwidget/src/main/res/values/colors.xml:6]: Oh, I understand the error in my thinking now. I was thinking the design system was implemented in AC but it's Fenix only. So, I would avoid that confusion by not using the layer_1 or primary text/icon naming convention.

Try something like `mozac_feature_searchwidget_text_color`

Comment (gabrielluong) [reply components/feature/searchwidget/src/main/res/values/colors.xml:6]: When we use this module in Fenix, I would like that we point `mozac_feature_searchwidget_text_color ` to `fx_mobile_text_color_primary` as an example so that we aren't falling back to the AC defined colors.

File: .buildconfig.yml
Comment (iorgamgabriel) [reply .buildconfig.yml]: Done

File: components/feature/searchwidget/src/main/res/values/colors.xml
Comment (iorgamgabriel) [reply components/feature/searchwidget/src/main/res/values/colors.xml]: Done

File: components/feature/searchwidget/src/main/java/mozilla/components/feature/searchwidget/AppSearchWidgetProvider.kt
Comment (iorgamgabriel) [reply components/feature/searchwidget/src/main/java/mozilla/components/feature/searchwidget/AppSearchWidgetProvider.kt]: Done

File: components/feature/searchwidget/src/main/res/values/dimens.xml
Comment (iorgamgabriel) [reply components/feature/searchwidget/src/main/res/values/dimens.xml]: Done

File: components/feature/searchwidget/src/main/res/values/strings.xml
Comment (iorgamgabriel) [reply components/feature/searchwidget/src/main/res/values/strings.xml]: Done

File: components/feature/search/src/main/res/values-night/colors.xml
Comment (gabrielluong) [inline components/feature/search/src/main/res/values-night/colors.xml:9]:
```suggestion
    <color name="mozac_feature_search_widget_text_color">@color/photonLightGrey05</color>
```

File: components/feature/search/src/main/res/values/colors.xml
Comment (gabrielluong) [inline components/feature/search/src/main/res/values/colors.xml:8]: I didn't see this being used anywhere on a quick search. Is this needed? I also saw we were search widget voice have a transparent background.

Comment (iorgamgabriel) [reply components/feature/search/src/main/res/values/colors.xml]: Yes they are user in mozac_search_widget_large.xml and mozac_search_widget_medium.xml . I added background transparent for Microphone button because will appear with grey background if I don't do this.

File: components/feature/search/src/main/res/values-night/colors.xml
Comment (iorgamgabriel) [reply components/feature/search/src/main/res/values-night/colors.xml]: Done

File: components/feature/search/src/main/AndroidManifest.xml
Comment (jonalmeida) [inline components/feature/search/src/main/AndroidManifest.xml:9]: I think I missed this earlier - why do we need the permission to query all packages (apps) installed on the user's phone? This permission is very powerful and requires us to provide a reason on how we use it when submitting an app to the Play store.

The warning `QueryAllPackagesPermission` below is meant to make us aware of what we are requesting, so if we do need this permission here, we should at the very least leave a comment on how it's used. 🙂

File: components/feature/search/src/main/java/mozilla/components/feature/search/widget/AppSearchWidgetProvider.kt
Comment (jonalmeida) [inline components/feature/search/src/main/java/mozilla/components/feature/search/widget/AppSearchWidgetProvider.kt:86]: nit: KDoc here please.

File: components/feature/search/src/main/java/mozilla/components/feature/search/widget/AppSearchWidgetProvider.kt
Comment (jonalmeida) [inline components/feature/search/src/main/java/mozilla/components/feature/search/widget/AppSearchWidgetProvider.kt:203]: Should this be in the config class too? R.string.app_name isn't a protected ID reference so apps might have even used `R.string.application_name` in their app and our overriding might not have worked.

**Comment by hneiva** [comment]
No RelEng items, removing review request

File: components/feature/search/src/main/AndroidManifest.xml
Comment (iorgamgabriel) [reply components/feature/search/src/main/AndroidManifest.xml:9]: This is needed for intentSpeech.resolveActivity from 
val intentSpeech = Intent(RecognizerIntent.ACTION_RECOGNIZE_SPEECH)

        return intentSpeech.resolveActivity(context.packageManager)?.let {
            PendingIntent.getActivity(
                context,
                REQUEST_CODE_VOICE, voiceIntent, PendingIntentUtils.defaultFlags
            )
        }
I can do this in manifest like in Focus and Fenix ,it's the same code or I can add this @SuppressLint("QueryPermissionsNeeded") to the method .

File: components/feature/search/src/main/java/mozilla/components/feature/search/widget/AppSearchWidgetProvider.kt
Comment (iorgamgabriel) [reply components/feature/search/src/main/java/mozilla/components/feature/search/widget/AppSearchWidgetProvider.kt:89]: Done

File: components/feature/search/src/main/java/mozilla/components/feature/search/widget/AppSearchWidgetProvider.kt
Comment (iorgamgabriel) [reply components/feature/search/src/main/java/mozilla/components/feature/search/widget/AppSearchWidgetProvider.kt]: Done

File: components/feature/search/src/main/java/mozilla/components/feature/search/widget/AppSearchWidgetProvider.kt
Comment (Mugurell) [inline components/feature/search/src/main/java/mozilla/components/feature/search/widget/AppSearchWidgetProvider.kt:242]: Looking at this being used in Fenix it wasn't immediately clear why this method needs [a new instance of `AppSearchWidgetProvider`](https://github.com/mozilla-mobile/fenix/pull/26331/files#diff-c9271216658683dbfa5f236fc4da16de698b708287062bd1af1767d2489e240cR94)
```kotlin
AppSearchWidgetProvider.updateAllWidgets(requireContext(), SearchWidgetProvider())
```
--- 

The easiest approach to avoid instantiating `SearchWidgetProvider` if we just need the class it to pass just that.
So this method could become
```kotlin
fun updateAllWidgets(context: Context, clazz: Class<AppSearchWidgetProvider>)
```
and then this would be called with
```kotlin
AppSearchWidgetProvider.updateAllWidgets(requireContext(), SearchWidgetProvider::class.java)
```
--- 

We could also make use of an inline type to have it declared here as
```kotlin
inline fun <reified T: KClass<out AppSearchWidgetProvider>> T.updateAllWidgets(context: Context)
```
and then this would be called with
```kotlin
SearchWidgetProvider::class.updateAllWidgets(requireContext())
```

File: components/feature/search/src/main/java/mozilla/components/feature/search/widget/VoiceSearchActivity.kt
Comment (Mugurell) [inline components/feature/search/src/main/java/mozilla/components/feature/search/widget/VoiceSearchActivity.kt:21]: Looking at how this is used on Fenix - where a bigger refactoring is needed to rename the current `VoiceSearchActivity` to `FenixVoiceSearchActivity` I propose having a different name in AC with maybe prepending `Base` or `Abstract` to `VoiceSearchActivity`, a naming scheme which is already used which would then allow to have only minimal changes in Fenix.

File: components/feature/search/src/main/java/mozilla/components/feature/search/widget/AppSearchWidgetProvider.kt
Comment (iorgamgabriel) [reply components/feature/search/src/main/java/mozilla/components/feature/search/widget/AppSearchWidgetProvider.kt]: Done

File: components/feature/search/src/main/java/mozilla/components/feature/search/widget/VoiceSearchActivity.kt
Comment (iorgamgabriel) [reply components/feature/search/src/main/java/mozilla/components/feature/search/widget/VoiceSearchActivity.kt]: Done

File: components/feature/search/src/main/java/mozilla/components/feature/search/widget/BaseVoiceSearchActivity.kt
Comment (Mugurell) [inline components/feature/search/src/main/java/mozilla/components/feature/search/widget/BaseVoiceSearchActivity.kt:70]: nit: This method forces children to record telemetry.
Allowing children to know of certain events and then deciding for themselves if they want to log telemetry or not seems more suitable.
eg: This could be:
```kotlin
fun onSpeechRecognitionStarted() = {}
```

File: components/feature/search/src/main/java/mozilla/components/feature/search/widget/BaseVoiceSearchActivity.kt
Comment (Mugurell) [inline components/feature/search/src/main/java/mozilla/components/feature/search/widget/BaseVoiceSearchActivity.kt:76]: nit: Continuing from the proposed method naming above this could become
```kotlin
abstract fun onSpeecRecognitionFinished(recognizedText: String?)
```

**Review by Mugurell — APPROVED** [review]
LGTM!
Seeing all previous comments were addressed I'll also give this a r+ but we should check with Jonathan for any other updates to be made before merging.

File: components/feature/search/src/main/java/mozilla/components/feature/search/widget/BaseVoiceSearchActivity.kt
Comment (iorgamgabriel) [reply components/feature/search/src/main/java/mozilla/components/feature/search/widget/BaseVoiceSearchActivity.kt]: Done

File: components/feature/search/src/main/java/mozilla/components/feature/search/widget/BaseVoiceSearchActivity.kt
Comment (iorgamgabriel) [reply components/feature/search/src/main/java/mozilla/components/feature/search/widget/BaseVoiceSearchActivity.kt]: Done

File: components/feature/search/src/main/java/mozilla/components/feature/search/widget/BaseVoiceSearchActivity.kt
Comment (jonalmeida) [inline components/feature/search/src/main/java/mozilla/components/feature/search/widget/BaseVoiceSearchActivity.kt:30]: Why do we create this result launcher in `onCreate` to only use it later on in the same callback method?

Also, you use the `@VisibleForTesting`, but I don't see any tests that use it.

Comment (iorgamgabriel) [reply components/feature/search/src/main/java/mozilla/components/feature/search/widget/BaseVoiceSearchActivity.kt]: Done

File: components/feature/search/src/main/java/mozilla/components/feature/search/widget/AppSearchWidgetProvider.kt
Comment (iorgamgabriel) [reply components/feature/search/src/main/java/mozilla/components/feature/search/widget/AppSearchWidgetProvider.kt]: Done

File: components/feature/search/src/main/AndroidManifest.xml
Comment (iorgamgabriel) [reply components/feature/search/src/main/AndroidManifest.xml]: This was needed for 
val intentSpeech = Intent(RecognizerIntent.ACTION_RECOGNIZE_SPEECH)
intentSpeech.resolveActivity(context.packageManager) 

to check if the phone has voice recognition.

File: components/feature/search/src/main/java/mozilla/components/feature/search/widget/BaseVoiceSearchActivity.kt
Comment (jonalmeida) [reply components/feature/search/src/main/java/mozilla/components/feature/search/widget/BaseVoiceSearchActivity.kt]: It's still unclear to me why we need to hold this in a var in this fashion. :)

**Review by jonalmeida — APPROVED** [review]
This looks good to me to land (after the logger lint is fixed). Great work! 🎉

File: components/feature/search/src/main/AndroidManifest.xml
Comment (jonalmeida) [reply components/feature/search/src/main/AndroidManifest.xml]: Left an offline comment to this discuss this since the permission was removed from the change and no longer a concern for this PR.

**Pull request merged**
